libnavui-voice/src/main/java/com/mapbox/navigation/ui/voice/api/MapboxSpeechLoader.kt

```diff
@@ -17,28 +20,88 @@ import com.mapbox.navigation.ui.utils.internal.resource.ResourceLoader
 import com.mapbox.navigation.ui.voice.model.TypeAndAnnouncement
 import com.mapbox.navigation.ui.voice.options.MapboxSpeechApiOptions
+import com.mapbox.navigation.utils.internal.InternalJobControlFactory
+import com.mapbox.navigation.utils.internal.logE
+import kotlinx.coroutines.cancel
+import kotlinx.coroutines.cancelChildren
+import kotlinx.coroutines.launch
 import java.net.MalformedURLException
 import java.net.URL
 
-internal class MapboxSpeechProvider(
+internal class MapboxSpeechLoader(
     private val accessToken: String,
     private val language: String,
     private val urlSkuTokenProvider: UrlSkuTokenProvider,
     private val options: MapboxSpeechApiOptions,
-    private val resourceLoader: ResourceLoader
-) {
+    private val resourceLoader: ResourceLoader,
+) : VoiceInstructionsDownloadTriggerObserver {
 
-    suspend fun load(typeAndAnnouncement: TypeAndAnnouncement): Expected<Throwable, ByteArray> {
+    private val currentRequests = mutableSetOf<Long>()
+    private val downloadedInstructions = mutableSetOf<TypeAndAnnouncement>()
+    private val downloadedInstructionsLock = Any()
+    private val defaultScope = InternalJobControlFactory.createDefaultScopeJobControl().scope
+
+    suspend fun load(voiceInstruction: VoiceInstructions): Expected<Throwable, ByteArray> {
         return runCatching {
-            val url = instructionUrl(typeAndAnnouncement.announcement, typeAndAnnouncement.type)
-            val response = resourceLoader.load(url)
+            val typeAndAnnouncement = VoiceInstructionsParser.parse(voiceInstruction)
+                .getValueOrElse { throw it }
+            val request = createRequest(typeAndAnnouncement).apply {
+                networkRestriction = NetworkRestriction.DISALLOW_ALL
+            }
+            val response = resourceLoader.load(request)
             return processResponse(response)
         }.getOrElse {
             createError(it)
         }
     }
 
-    private suspend fun ResourceLoader.load(url: String) = load(ResourceLoadRequest(url))
+    override fun trigger(voiceInstructions: List<VoiceInstructions>) {
+        defaultScope.launch {
+            voiceInstructions.forEach { voiceInstruction ->
+                val typeAndAnnouncement = VoiceInstructionsParser.parse(voiceInstruction).value
+                if (typeAndAnnouncement != null && !hasTypeAndAnnouncement(typeAndAnnouncement)) {
+                    predownload(typeAndAnnouncement)
+                }
+            }
+        }
+    }
+
+    fun cancel() {
+        defaultScope.cancel()
+        currentRequests.forEach { resourceLoader.cancel(it) }
+    }
+
+    private fun hasTypeAndAnnouncement(typeAndAnnouncement: TypeAndAnnouncement): Boolean {
+        synchronized(downloadedInstructionsLock) {
+            return typeAndAnnouncement in downloadedInstructions
+        }
+    }
+
+    private fun predownload(typeAndAnnouncement: TypeAndAnnouncement) {
```

Comment:
How does the system cleanup unplayed voice instructions?

Reply:
The same way it cleans up played instructions. :)
We have a mechanism of cleaning them but we only clean the files that were created by the Nav SDK during `generate` method. 
I'll ask Core SDK but will it change anything? I mean the only difference with what we have now on main branch is that for un unfinished route there will be slightly more instructions in the cache.

Reply:
Nice, generate will clean up played instructions. 

Is it possible for the preloading instructions, to cause disk or out of memory issues? For example, the route preloads 10GB of instructions but only 1GB of of space is available (exaggerating actual space values)

Reply:
> created by the Nav SDK during generate method.

Also curious, unplayed instructions will only be cleared when `generate` is called? generate may not be called, right?

Reply:
`generate` will not clean up played instructions. It will **create** a file (see [this line](https://github.com/mapbox/mapbox-navigation-android/blob/8248c1ffc4878133535ca2b37ff8c3256bb53d55/libnavui-voice/src/main/java/com/mapbox/navigation/ui/voice/api/MapboxVoiceApi.kt#L27)) that will be cleaned by user's callback that they pass to `MapboxVoiceInstructionsPlayer#play`. Predownloading does not pass through this line, therefore this file will not be created.
However, there is also cache handled by the Core SDK. And it is not cleared by us. Neither now, nor on main. And there is no way right now to trigger some specific file removal. Capturing from a Slack thread:

> Resource cache is not cleared explicitly. TileStore has a unified quota for tiles and resources and will evict individual items when the quota is exceeded. Resources (including tiles) with the closest expiration date are removed until the quota is not exceeded anymore.
A mechanism to clear the cache might be part of the OP plan for 2023.
...
By default the quota is unlimited, but there is logic to evict resources if the disk space is low. So for removal to happen, either a quota must have been set an exceeded or the disk space must be running low.